api/service/api.go

```diff
@@ -0,0 +1,240 @@
+		return common.HexToHash(id), nil
+	} else if isSlot(id) || isKnownIdentifier(id) {
+		a.metrics.RecordBlockIdType(m.BlockIdTypeBeacon)
+		result, err := a.beaconClient.BeaconBlockHeader(context.Background(), &api.BeaconBlockHeaderOpts{
+			Common: api.CommonOpts{},
+			Block:  id,
+		})
+
+		if err != nil {
+			var apiErr *api.Error
+			if errors.As(err, &apiErr) {
+				switch apiErr.StatusCode {
+				case 404:
+					return common.Hash{}, errUnknownBlock
+				}
+			}
+
+			return common.Hash{}, errServerError
+		}
+
+		return common.Hash(result.Data.Root), nil
+	} else {
+		a.metrics.RecordBlockIdType(m.BlockIdTypeInvalid)
+		return common.Hash{}, newBlockIdError(id)
+	}
+}
+
+// blobSidecarHandler implements the /eth/v1/beacon/blob_sidecars/{id} endpoint, using the underlying DataStoreReader
+// to fetch blobs instead of the beacon node. This allows clients to fetch expired blobs.
+func (a *API) blobSidecarHandler(w http.ResponseWriter, r *http.Request) {
+	param := chi.URLParam(r, "id")
+	beaconBlockHash, err := a.toBeaconBlockHash(param)
+	if err != nil {
+		err.write(w)
+		return
+	}
+
+	result, storageErr := a.dataStoreClient.Read(r.Context(), beaconBlockHash)
+	if storageErr != nil {
+		if errors.Is(storageErr, storage.ErrNotFound) {
+			errUnknownBlock.write(w)
+		} else {
+			a.logger.Info("unexpected error fetching blobs", "err", storageErr, "beaconBlockHash", beaconBlockHash.String(), "param", param)
+			errServerError.write(w)
+		}
+		return
+	}
+
+	blobSidecars := result.BlobSidecars
+
+	filteredBlobSidecars, err := filterBlobs(blobSidecars.Data, r.URL.Query().Get("indices"))
+	if err != nil {
+		err.write(w)
+		return
+	}
+
+	blobSidecars.Data = filteredBlobSidecars
+	responseType := r.Header.Get("Accept")
+
+	if responseType == sszAcceptType {
+		res, err := blobSidecars.MarshalSSZ()
+		if err != nil {
+			a.logger.Error("unable to marshal blob sidecars to SSZ", "err", err)
+			errServerError.write(w)
+			return
+		}
+
+		_, err = w.Write(res)
+
+		if err != nil {
+			a.logger.Error("unable to write ssz response", "err", err)
+			errServerError.write(w)
+			return
+		}
+	} else {
+		err := json.NewEncoder(w).Encode(blobSidecars)
+		if err != nil {
+			a.logger.Error("unable to encode blob sidecars to JSON", "err", err)
+			errServerError.write(w)
+			return
+		}
+	}
+}
+
+// filterBlobs filters the blobs based on the indices query provided.
+// If no indices or invalid indices are provided, the original blobs are returned.
+func filterBlobs(blobs []*deneb.BlobSidecar, indices string) ([]*deneb.BlobSidecar, *httpError) {
+	if indices == "" {
+		return blobs, nil
+	}
+
+	splits := strings.Split(indices, ",")
+	if len(splits) == 0 {
+		return blobs, nil
+	}
+
+	indicesMap := map[deneb.BlobIndex]struct{}{}
+	for _, index := range splits {
+		parsedInt, err := strconv.ParseUint(index, 10, 64)
+		if err != nil {
+			return nil, newIndicesError(index)
+		}
+		blobIndex := deneb.BlobIndex(parsedInt)
```

Comment:
would return error here if the blobIndex is out of range.

Reply:
My intention here is to mirror the CL client behavior as close as possible, if the index is out of range/duplicated etc it's just ignored (at least on Lighthouse)

```
❯ curl -s "<cl-url>/eth/v1/beacon/blob_sidecars/0x8f0d0ca5a16fc064514574dddde8d11b39e81f29b642d9f8b149a369b7c9fa9e?indices=10"
{"data":[]}%
```

```
❯ curl -s "<cl-url>/eth/v1/beacon/blob_sidecars/0x8f0d0ca5a16fc064514574dddde8d11b39e81f29b642d9f8b149a369b7c9fa9e?indices=1,1,10"
#.. returns a single blob at index 1
```

Reply:
is that in the spec or is that just what Lighthouse does?  Duplicated I get, out of range seems worth alerting the caller over.

Reply:
It's the behavior or lighthouse, the [spec](https://github.com/ethereum/beacon-APIs/blob/master/apis/beacon/blob_sidecars/blob_sidecars.yaml#L1) doesn't specify :D (it's very possible there's a more detailed spec I'm missing though).